File: Julia/DRIP.jl
## A Software Package for Solving Dynamic Rational Inattention Problems (D.R.I.P.)
## Copyright © 2020 Hassan Afrouzi, Choongryul Yang and Miguel Acosta
## When using this package for your work, cite Afrouzi and Yang (2019) for reference.
## Julia code written by Miguel Acosta / Modified by Hassan Afrouzi

using LinearAlgebra

## A General Structure for D.R.I.P.
struct drip 
    ω; β; A; Q; H;                    # primitives
    K; Y; Σ_z; Σ_p; Σ_1; Ω;           # solution
end

## A General Structure for Transition dynamics of Rational Inattention Problems (T.R.I.P.)
struct trip 
    P::drip;                # problem and solution in steady state
    T::Int;                 # length of T.R.I.P.
    Σ_1s; Σ_ps; Ωs;         # priors, posteriors and benefit matrices 
    Ds;                     # eigenvalues and eigenvectors of Σ_t^(0.5)Ω_tΣ_t^(0.5)
end 

## A Structure for the impulse responses of D.R.I.P. for the steady state posterior
struct dripirfs
    T     :: Int            # length of IRFs
    x     :: Array{Float64} # IRFs of the fundamental shocks
    x_hat :: Array{Float64} # IRFs of beliefs
    a     :: Array{Float64} # IRFs of actions
end

## solve_drip: this function solves for the steady state of of the D.R.I.P.
## Inputs
# ω      : Cost of information
# β      : Discount factor
# A      : Transition matrix: x=Ax+Qu
# Q      : Std. Dev. matrix: x=Ax+Qu
# H      : Mapping of shocks to actions: v=-0.5(a'-x'H)(a-H'x)

## Outputs: a drip structure with 
# Σ_1    : Steady-state Prior uncertainty
# Σ_p    : Steady-state posterior uncertainty
# Λ      : Shadow matrix on the no-forgetting constraint -- Λ*(Σ-Σ_p) = 0
# Ω      : Dynamic benefit matrix
# Y      : Weight vector for evolution of actions
# Σ_z    : Covariance matrix of the rational inattention error
# K      : Kalman gain matrix

function solve_drip(ω,β,A,Q,H;              # primitives of the D.R.I.P.
                    fcap::Bool = false,     # optional: if true then solves the problem with fixed capacity κ = ω.
                    Ω0         = H*H',      # optional: initial guess for steady state information matrix
                    Σ0         = A*A'+Q*Q', # optional: initial guess for steady state prior
                    w          = 1,         # optional: updating weight in iteration
                    tol        = 1e-4,      # optional: tolerance level for convergence
                    maxit      = 10000)     # optional: maximum number of iterations

    ## initialize
    (n,m) = length(size(H)) == 2 ? size(H) : (size(H,1),1);  
    # n: dimension of state, m: number of actions
    eye   = Matrix{Float64}(I,n,n);
    err   = 1
    iter  = 0
    SqRΣ  = sqrt(Σ0)
    Ω_c   = H*H';
    Σ1    = Matrix{Float64}(I,n,n)
    Σ_p   = Matrix{Float64}(I,n,n)
    Λ     = Matrix{Float64}(I,n,n)
    κ     = ω
    # iterate 
    while (err > tol) & (iter < maxit)
        D, U    = eigen(SqRΣ*Ω0*SqRΣ);
        D       = diagm(getreal(D));
        U       = getreal(U);
        
        if fcap == true 
            ω = (2^(2*κ)/det(max.(ω*eye,D)))^(-1/n);
        end
        Λ       = U*(max.(ω*eye-D,0.0))*U';
        Σ_p     = getreal(ω*SqRΣ*U/(max.(D,ω*eye))*U'*SqRΣ);

        Σ1      = A*Σ_p*A' + Q*Q'
        err     = norm(Σ1 - Σ0,2)/norm(Σ0,2);

        Σ0      = w*Σ1 + (1-w)*Σ0

        SqRΣ    = sqrt(Σ0);
        invSqRΣ = pinv(SqRΣ);

        Ω0      = w*(Ω_c + β*A'*invSqRΣ*(ω*eye - Λ)*invSqRΣ*A)+(1-w)*Ω0;
        Ω0      = (abs.(Ω0).>1e-10).*Ω0;

        iter   += 1
    end

    if iter == maxit
        print("RI Code hit maxit\n")
    end

    Σ1     = (abs.(Σ1).>1e-10).*Σ1 + diagm(abs.(diag(Σ1)).<= 1e-10)*1e-8
    inv_Σ1 = pinv(Σ1) ;
    Y      = (eye - Σ_p*inv_Σ1)'*H ;
    Σ_z    = H'*(Σ_p - Σ_p*inv_Σ1*Σ_p)*H ;
    K      = Σ1*Y*pinv(Y'*Σ1*Y + Σ_z) ;

    return(drip(ω,β,A,Q,H,K,Y,Σ_z,Σ_p,Σ1,Ω0))
end

## solve_trip: this function solves for the transition dynamics given an initial prior

function solve_trip(P::drip,         # D.R.I.P.
                    Σ0;              # initial prior matrix
                    T     = 100,     # optional: time until convergence to steady state
                    tol   = 1e-4,    # optional: tolerance for convergence
                    maxit = 1000     # optional: max iterations
                    )
    ## Initialize 
    Ωs  = repeat(P.Ω, inner = [1,1,T]);
    Ωsp = Ωs;

    Σ_1s  = repeat(P.Σ_1, inner = [1,1,T]); Σ_1s[:,:,1] = Σ0;
    Σ_1sp = Σ_1s;
    (n,m) = length(size(P.H)) == 2 ? size(P.H) : (size(H,1),1);  
    # n: dimension of state, m: number of actions

    Σ_ps= repeat(P.Σ_p, inner = [1,1,T]); # initialize posteriors
    Ds  = zeros(n,T);                          # initialize eigenvalues 

    iter = 0;
    err  = 1;
    eye  = Matrix(I,n,n);
    while (err > tol) & (iter <= maxit)
        ## Given Ωs, find Sigmas using the law of motion for priors 
        for i in 1:1:T-1
            SqSigma        = real.(sqrt(Σ_1s[:,:,i]));
            D, U           = eigen(SqSigma*Ωs[:,:,i]*SqSigma);
            Ds[:,i]        = real.(D);
            U              = real.(U);
            D              = diagm(Ds[:,i]);
            Σ_ps[:,:,i]    = P.ω*SqSigma*U/(max.(D,P.ω*eye))*U'*SqSigma;
            Σ_1sp[:,:,i+1] = P.Q*P.Q'+P.A*Σ_ps[:,:,i]*P.A';
        end
        # Given Σ_1s, Find Omegas using the Euler equation
        for i = T-1:-1:1
            SqSigma      = sqrt(Σ_1sp[:,:,i+1]);
            invSqSigma   = real.(pinv(SqSigma));
            D, U         = eigen(SqSigma*Ωs[:,:,i+1]*SqSigma);
            D            = diagm(getreal(D));
            U            = real.(U);
            Ωsp[:,:,i]   = P.H*P.H'+P.β*P.A'*invSqSigma*U*min.(D,P.ω*eye)*U'*invSqSigma*P.A;
        end
        err    = norm(Σ_1sp-Σ_1s)/norm(Σ_1s);
        Σ_1s   = real.(Σ_1sp);
        Ωs     = real.(Ωsp);
        iter  += 1;
    end
    # Throw error if T was too short for convergence to steady state
    con_err = norm(P.Q*P.Q'+P.A*Σ_ps[:,:,end-1]*P.A'-P.Σ_1)/norm(P.Σ_1);
    if (con_err  > tol)
        error("T was too short for convergence. Try larger T.");
    end 
    # Finally, store the eigenvalues of steady state
    SqSigma      = real.(sqrt(Σ_1s[:,:,end]));
    D, U         = eigen(SqSigma*Ωs[:,:,end]*SqSigma);
    Ds[:,end]    = real.(D);
    # return the trip structure 
    return(trip(P,T,Σ_1s,Σ_ps,Ωs,Ds))
end

########## Aux. Functions ############

function getreal(M)
    #if maximum(abs.(imag.(M))) > 1e-6
    #    print("Warning: Matrix decomposition returned complex numbers larger than 1e-6")
    #end
    return(real.(M))
end

function infinitesum(func; tol = 1e-6,maxit = 1000,start=0)
    diff  = 1.0
    infsum = func(start)
    it    = start + 1
    while (diff > tol) & (it < maxit)
        func_it = func(it)
        infsum += func_it
        diff = maximum(func_it)
        it += 1
    end
    return(infsum)
end

## calculate the amount of information acquired in bits/nats for a drip 
function capacity(P::drip;      # drip structure
                  unit = "bit"  # optional: unit of capacity (bit or nat). 
                  )
    if unit == "bit"
        κ = 0.5*log(det(P.Σ_1)/det(P.Σ_p))/log(2); #returns capacity in bits
    elseif unit == "nat" 
        κ = 0.5*log(det(P.Σ_1)/det(P.Σ_p));        #returns capacity in nats 
    else 
        println("Invalid input for unit! Capacity is reported in bits.")
        κ = 0.5*log(det(P.Σ_1)/det(P.Σ_p))/log(2);
    end
    return(κ)
end

function dripirfs(P::drip,T::Int)
    (n,m) = length(size(P.H)) == 2 ? size(P.H) : (size(P.H,1),1)
    (_,k) = length(size(P.Q)) == 2 ? size(P.Q) : (size(P.Q,1),1)
    x     = zeros(n,k,T);
    x_hat = zeros(n,k,T);
    a     = zeros(m,k,T);
    for kk in 1:k
        e_k = zeros(k,1); e_k[kk] = 1;
        for ii in 1:T
            if ii==1 
                x[:,kk,ii]     = P.Q*e_k;
                x_hat[:,kk,ii] = (P.K*P.Y')*(x[:,kk,ii]);
            else 
                x[:,kk,ii]     =P.A*x[:,kk,ii-1];
                x_hat[:,kk,ii] =P.A*x_hat[:,kk,ii-1]+(P.K*P.Y')*(x[:,kk,ii]-P.A*x_hat[:,kk,ii-1]);
            end
            a[:,kk,ii]  .= P.H'*x_hat[:,kk,ii];
        end
    end
    return(dripirfs(T,x,x_hat,a))
end

function dripirfs(Pt::trip,T::Int)
    Ss    = Pt.P; # steady state drip
    (n,m) = length(size(Ss.H)) == 2 ? size(Ss.H) : (size(Ss.H,1),1) # dimensions of state and actions
    (_,k) = length(size(Ss.Q)) == 2 ? size(Ss.Q) : (size(Ss.Q,1),1) # number of structural shocks
    eye   = Matrix(I,n,n);

    x     = zeros(n,k,T); # initialize IRFs of state 
    x_hat = zeros(n,k,T); # initialize IRFs of beliefs 
    a     = zeros(m,k,T); # initialize IRFs of actions
    for kk in 1:k
        e_k = zeros(k,1); e_k[kk] = 1;
        for ii in 1:T
            if ii==1 
                x[:,kk,ii]     = Ss.Q*e_k;
                x_hat[:,kk,ii] = (eye-Pt.Σ_ps[:,:,ii]*pinv(Pt.Σ_1s[:,:,ii]))*(x[:,kk,ii]);
            else 
                x[:,kk,ii]     =Ss.A*x[:,kk,ii-1];
                x_hat[:,kk,ii] =Ss.A*x_hat[:,kk,ii-1]+(eye-Pt.Σ_ps[:,:,ii]*pinv(Pt.Σ_1s[:,:,ii]))*(x[:,kk,ii]-Ss.A*x_hat[:,kk,ii-1]);
            end
            a[:,kk,ii]  .= Ss.H'*x_hat[:,kk,ii];
        end
    end
    return(dripirfs(T,x,x_hat,a))
end

File: Matlab/src/Drip.m
%% Afrouzi and Yang (2019)
%% This code solves for the steady state of the dynamic 
%% multivariate rational inattention problem in LQG setting.
%
%     Drip(omega,beta,A,Q,H, kwargs...) -> Drip
%
% Solves for the steady state of a Dynamic Rational Inattention Problem defined 
% by the arguments and stores the solution in a Drip type.
% See [Afrouzi and  Yang (2020)](http://afrouzi.com/dynamic_inattention.pdf) for details.
%
%% ARGUMENSTS
%
% The function takes the primitives of the Drip as arguments:
%     * omega  : Cost of information
%     * beta   : Discount factor
%     * A      : Transition matrix: x=Ax+Qu
%     * Q      : Std. Dev. matrix: x=Ax+Qu
%     * H      : Mapping of shocks to actions: v=-0.5(a'-x'H)(a-H'x)
%
%% OPTIONAL ARGUMENSTS (kwargs...)
%
% Default values are set unless specified otherwise by user.
%     * fcap      [= false]    [if `true` then solves the problem with fixed capacity = ω bits]
%     * initOmega [= H*H']     [initial guess for steady state information matrix]
%     * initSigma [= A*A'+Q*Q'][initial guess for steady state prior]
%     * w         [= 1]        [updating weight on the new guess in iteration]
%     * tol       [= 1e-4]     [tolerance level for convergence]
%     * maxit     [= 10000]    [maximum number of iterations]
%
%% OUTPUTS
%  The function returns a structure with the primitives and the solution objects:
%     * ss : a structure with the solution to the steady state of the DRIP with the following fields:
%           * Y       : Weight vector for evolution of actions
%           * Sigma_z : Covariance matrix of the rational inattention error
%           * K       : Kalman gain matrix
%           * Sigma_1 : Steady-state prior covariance matrix under the solution
%           * Sigma_p : Steady-state posterior covariance matrix under the solution
%           * Omega   : Dynamic benefit matrix
%     * also contains fields for the primitives of the problem (omega,beta,A,Q,H)
%
%% EXAMPLE 
% >> p = Drip(omega,beta,A,Q,H)
% 

function p = Drip(omega,beta,A,Q,H,varargin) 

    % parse optional inputs
    args = inputParser;
    addOptional(args,'initOmega',H*H');
    addOptional(args,'initSigma',A*A'+Q*Q');
    addOptional(args,'tol',1e-4);
    addOptional(args,'fcap',false,@islogical);
    addOptional(args,'w',1);
    addOptional(args,'maxit',10000);

    parse(args,varargin{:});

    % initialize 

    [n,~]      = size(H)   ;                % get the dimension of the state n 
    p.ss.err   = 1 ;                        % initialize convergence error
    iter       = 0 ;                        % initialize number of iterations
    Sigma0     = args.Results.initSigma ;   % initial guess for Omega
    p.ss.Omega = args.Results.initOmega ;   % initial guess for Sigma
    
    % other variables to be used in the code:
    SqRSigma   = sqrtm(Sigma0);
    Omega_c    = H*H'      ;
    I          = eye(n)    ;

    % if the user wants solution with fixed capacity, set kappa = omega
    if args.Results.fcap == true 
        kappa     = omega ;
    end 	
    
    % Loop (stop if converged or iteration reached maxit)
    while (p.ss.err > args.Results.tol) && (iter <= args.Results.maxit)
		
        % get the eigenvalue and eigen vectors of SqSigma*Omega*SqSigma
		X     = SqRSigma*p.ss.Omega*SqRSigma ;
        [U,D] = eig(X) ;
        D = real(D);
        U = real(U);
        
        % if the user wants solution with fixed capacity, find the
        % corresponding lagrange multiplier on the capacity constriant:
        if args.Results.fcap == true
            omega = (2^(2*kappa)/det(max(D,omega*I)))^(-1/n);
        end
        
        % use the policy function in Afrouzi and Yang to get the implied
        % posterior covariance:
        p.ss.Sigma_p = omega*SqRSigma*U/(max(D,omega*I))*U'*SqRSigma;
    
        % get the new guess for prior covariance and calculate the
        % convergence error:
        p.ss.Sigma_1 = A*p.ss.Sigma_p*A' + Q*Q' ;
        p.ss.err     = norm(p.ss.Sigma_1 - Sigma0)/norm(Sigma0);

        % update the guess for prior covariance given update weight w
        Sigma0      = args.Results.w*p.ss.Sigma_1 + (1-args.Results.w)*Sigma0 ;
        
        % use the Euler equation in Afrouzi and Yang to update guess for
        % Omega given update weight w:
        SqRSigma    = real(sqrtm(Sigma0));
        invSqRSigma = real(pinv(SqRSigma));
        p.ss.Omega   = args.Results.w ...
                        *(Omega_c + beta*A'*invSqRSigma*U*(min(D,omega*I))*U'*invSqRSigma*A) ...
                        +(1-args.Results.w)*p.ss.Omega;
        p.ss.Omega = (abs(p.ss.Omega)>1e-10).*p.ss.Omega ;

        % update number of iterations
        iter        = iter + 1;
    end
    
    % produce matrices Y, K, Sigma_z as in Afrouzi and Yang for output
    inv_Sigma1  = pinv(p.ss.Sigma_1) ;
    p.ss.Y       = (I - p.ss.Sigma_p*inv_Sigma1)'*H ;
    p.ss.Sigma_z = H'*(p.ss.Sigma_p - p.ss.Sigma_p*inv_Sigma1*p.ss.Sigma_p)*H ;
    p.ss.K       = p.ss.Sigma_1*p.ss.Y*pinv(p.ss.Y'*p.ss.Sigma_1*p.ss.Y + p.ss.Sigma_z) ;
    
    % store the primitives of the problem
    p.omega = omega;
    p.beta = beta;
    p.A = A;
    p.Q = Q;
    p.H = H;
end
